### ISSUE TITLE
Computing each tile from 2 sources on the fly. 

### ISSUE THREAD

**Issue by Medwar** [openseadragon/openseadragon#1371]
Hi! 
Could you please advise, what is the best solution to make tile computed and drawn based on two sources? Like filters do, with the difference that there will be two sources. I have searched through the issues and found a few possible solutions, #126, #765, but they are from 2015 or earlier, so I decided to ask - maybe an easier and more elegant way has already appeared in OSD.  

So far the way that I see, is to make a second, invisible instance of OSD, try to catch the 'tile-loaded' event from both, and compute the resulting tile.
Perhaps there may be a way to make this with "World", but I don't see how.

**Comment by iangilman** [comment]
Interesting question! I don't think anyone has done this directly (it seems like it could be a great plug-in). The technique you mention seems valid, but like you say, we can now do it in a single viewer with the "world". Open both images and make sure they are at the same place in the viewer (they will be by default if you don't specify, and if they have the same aspect ratio). Set one of the images to opacity 0 and preload true, and you should be able to grab its tile-loaded events.

**Comment by Medwar** [comment]
Hi Ian! The solution is working. Draft demo is on codepen, link below. Not everything is tested yet, and there are restrictions: sources must be of the same size, there should be only one merging operation for a viewer. 
The number of source images to be merged is not limited however. 
Nice point is that filters from filtering plugin will be applied to each source before merging. But only before. 
https://codepen.io/Medwar/pen/godENJ?editors=0010
(referred to https://codepen.io/Medwar/pen/MrqRWp?editors=0010)

I took the OpenSeadragon filtering plugin, and reworked the methods.
I deleted the `'tile-drawing'` handler, 
because the plugin seems to be working only with the `'tile-loaded'` handler,
if I write the resulting image into `tile.context2D`

But there are two problems. 
1) One is with Cross-domain tainted canvas. 
https://blog.codepen.io/2013/10/08/cross-domain-images-tainted-canvas/
I failed to solve this for my hosting, don't know what I am doing wrong. 
At the hosting, I put into .htaccess of the images folder the following lines:
`
header unset access-control-allow-origin 
header set access-control-allow-origin * 
`
I found this instrument https://www.test-cors.org/ for checking CORS policy on a remote server, but don't exactly how to interpret what I got.
I finished with installing on my PC a chrome plugin that disables CORS security for certain addresses. After that for me everything works fine.
But I cannot share my codepen, because anyone else will be getting CORS errors.

2) The second problem that after I removed the `'tile-drawing'` handler from the filtering plugin,
   it doesn't redraw the initial canvas content after resetting it when the filters are enabled.
   it does the redraw only if an action is taken, pan or zoom, even if only for 1px distance.
   I tried to solve it. Please see the comments in https://codepen.io/Medwar/pen/MrqRWp?editors=0010  near
`
for (var i = 0; i < itemsToReset.length; i++) {
	itemsToReset[i].reset(); 
}
`
		   
Could you please have a look?

**Comment by Medwar** [comment]
Also, how to insert code here with keeping the line breaks? `It's inserted 
as one line`

**Comment by Medwar** [comment]
I want to develop stereo-image viewing functionality with OpenSeadragon. 

Above, merging the sterepair of 2 images to anaglyph on the fly,  for Red-Cyan or Green-Magenta 3D glasses,  is already working.
 
Second, interlaced view for 3D-monitors is working in my second pen: https://codepen.io/Medwar/pen/KZrorO?editors=0010
Here I hacked the `blendSketch` method by introducing a new `compositeOperation` value and processing it in the code directly. Perhaps a custom `compositeOperation` set up with a user function could be made here, but I see no application for it except my interlace feature. For interlace functionality I think it's the best place, as it has to be done on finally drawn canvas without scaling to exactly fit the monitor's pixels, and interlace itself is a sort of compositeOperation.
 
Third, side-by-side view, I am still working on it.

I don't know how to do better, leave this stuff as a set of hack-solutions, 
or make a full plugin out of it (with my current skill it may appear a hard task). 
If making a plugin many questions arise: how not to override the standard methods? E.g. with interlace. It is bad practice as far as I understand. Or, if I am forking and seriously changing the filtering plugin, and adding to it an ANAGLYPH function, rewritten, but in source taken from another open GitHub project what should I wright in the copyright?
I am new to gitHub and this is my first experience with JS, so sorry that I don't know common best practices yet.

**Comment by iangilman** [comment]
Awesome! Sounds like you've made excellent progress! I'll try to answer your questions; let me know if I missed some.

Unfortunately I don't know much about CORS, but there must be some way to make it work.

If you need to force a render after resetting, you can call `viewer.forceRedraw()`.

To insert code here with line breaks, you use ``` like so:

```
hello!
```

Regarding overriding standard methods, the best strategy is to update OSD to take in a callback or fire an event or something that allows for customization. In other words, wherever you can't accomplish what you need to without overriding a method, that method needs to be fixed to be able to support the addition of your customization. Sounds like the addition of `compositeOperation` in `blendSketch` is a great example.

Regarding the license, one consideration is how much of the original code you're using and how much new code you're adding. If you're just adding a little code, you'd typically just keep the original license. If it's mostly new code, maybe you would say "some portions licensed X from Y project" or maybe keep them in their own file with their own license. Of course, if your changes make sense as extensions to the original code, the author might be happy to have them added to their project. If your extension move beyond the scope of the original plugin (which I think is the case here) you're better off making a new project (but acknowledging where the code came from).

In the case of the filtering plugin, it's an unusual license: 

https://github.com/usnistgov/OpenSeadragonFiltering/blob/master/LICENSE

Assuming you're using their code but have added significant portions of your own, I would say you should pick a new open source license (MIT is common, for instance, and OSD uses BSD) and include in your LICENSE file something about "some portions" and the filtering license and a link to the filtering plugin. 

@avandecreme any thoughts on the licensing front?

@medwar it would be awesome if you could create a plugin (or multiple if appropriate) with your code! I think it shouldn't be too hard, and I'm happy to help point you in the right direction. Here's our little writeup on getting started:

https://github.com/openseadragon/openseadragon/wiki/Making-OpenSeadragon-Plugins

**Comment by avandecreme** [comment]
Yes, you are free to do anything you want with the filtering plugin code since it is public domain.

**Comment by Medwar** [comment]
Thank you! 
I will fix the CORS problem and will return here with working demos & questions. After some time.

The second problem with not-redrawing, sorry, it was my own bug. Fixed.
In some cases I did not pass the original event's callback to the filter function to call in the end, and that was the issue. That stopped all the further OSD logic including redrawing. Don't understand yet how to easily debug the issues with callbacks. Not obvious.

**Comment by iangilman** [comment]
@avandecreme Great! It seems like it would be nice to reference the original, but it sounds like it's not strictly necessary? 

@Medwar sounds good... Let us know if you need help along the way!

**Comment by Medwar** [comment]
1) CORS - still fighting. I have difficulties as I have only shared hosting with very restricted administration rights.  
Is it possible to place my 2 demo tiled images on the same resourse on which "duomo" tiled image is hosted? With "duomo" all pens work ok, but merging must be demonstrated on two same-sized images.

2) Hook into "Drawer.blendSketch"
2.1)  Am I thinking the right way? Is it a worthwhile solution: 
to change the main code to give ability to pass a custom external function into `'customOperation'` option of the OSD?

I can offer such a change of the main code:
`Drawer.blendSketch`
```
 if (compositeOperation) {
            this.context.globalCompositeOperation = compositeOperation;
        }
```
to change for:
```
        if (compositeOperation && (typeof(compositeOperation) != 'function')) {
            this.context.globalCompositeOperation = compositeOperation;
        }
```

And add a following condition and run the function, and the similar into another branch, where bounds are not set.
```
            if (typeof(compositeOperation) != 'function'){ 
                // normal flow  
                ...
                );
            } else {  // compositeOperation is a function
                // run custom 'compositeOperation' function 
                compositeOperation(this, bounds, function(){});
            }  
```

Tested here:
https://codepen.io/Medwar/pen/Ozdova?editors=0010
  (yet without code for (!bounds) - branch where the sketchcanvas is scaled).

Perhaps the other way may be emitting a new event, but does it worth it?


2.2) Doing so, in this method - `Drawer.blendSketch` - we can theoretically give possibility to merge two images in any custom way, e.g. create anaglyph and so on.
For generating interlaced image this place is most suitable in my opinion (cause for interlacing we need to bind with display's pixel lines, no scaling). 
However for slower filters it appears to be a really bad place. 
Just as an experiment I put in the 'compositeOperation' the slightly modified ANAGLYPH function from my "Filtering&Merging" solution. And it works, but with huge lags.
I see that `Drawer.blendSketch` is called very often, once for every drawing step when image is drawn from blurry to sharp, and once for every step of animation. 
In the test pen above I inserted console.log("interlace, anaglyph") to monitor the frequency of calls. 
Is there a possibility to reduce the number of animation steps => the number of calls to .blendSketch ?

In fact my own application needs only interlace at this step, but I am digging deeper to create a more general, universal hook here. Most "merged" images that I can think of can be genereted on the fly in 'tiles-loaded'. 'blendSketch' may be suitable for some complex merging where the whole visible part of the image should be read by merging algorythm. 

3) "Filtering&Merging" - still working on it.  Most possibly it will be a plugin - replacement  for "OpenSeadragon-filtering" (like new version of it), keeping the functionality of the original, but adding ability to use filters consuming several source `tiledImage`s.

**Comment by iangilman** [comment]
> Is it possible to place my 2 demo tiled images on the same resourse on which "duomo" tiled image is hosted? With "duomo" all pens work ok, but merging must be demonstrated on two same-sized images.

Duomo is just stored in GitHub pages (under the OpenSeadragon organization). You can do the same with your GitHub account: 

https://pages.github.com/

>  Am I thinking the right way? Is it a worthwhile solution:
to change the main code to give ability to pass a custom external function into 'customOperation' option of the OSD?

I like the idea of the custom method, but I don't think we should overload `compositeOperation`. Right now it is clearly just for the strings normally accepted by canvas. Adding a function possibility to the same parameter seems like it'll cause confusion, and we already have areas like that in the code that do cause confusion. So, I'm in favor of adding a new parameter that has the custom method.

As for whether it should be an event versus a custom parameter, I guess it depends on whether we want there to be multiple of them (which an event would support). I imagine just one function at a time is good, so we probably don't need it to be an event. Also the custom function is probably a little better for performance, since we don't need to fire it speculatively.  

There's no way to call blendSketch less frequently, so I agree it should only be used for things that really need it. Otherwise doing the compositing in `tiled-loaded` is the right choice.

**Comment by Aiosa** [comment]
I actually implemented this with webgl. I am using modified IIPImage server to fetch any number of images (which is a bit of compatibility drag).

I was actually thinking about enabling TiledImage : Tile  relationship as N:M instead of current N :1 (now tiles can belong to multiple sources thanks to cache, what about source having multiple tiles?). I was struggling with the need for advanced server that supports synchronized, tile array requests. This also prevents me from using the cache at 100% since when data changes I have to re-fetch all of the tiles. If tile source had them separated (no matter whether tiles are fetched all at once or separately), I could have tiled image owning tiles with dynamic count, use cache full potential and have flexible and rather powerful way of optimizing the fetching... @iangilman what do you think about this feature?

**Comment by iangilman** [comment]
So the goal would be that for any one tile, it could potentially have source from more than one location? Might it be possible to do that already with the changes you made to the tile loading? Instead of loading a single thing, you would load several things and only return once they had all arrived?

**Comment by Aiosa** [comment]
One approach would be to let the TileSource download multiple items (or let TiledImage own multiple TileSources), either synchronously (no or limited changes), or asynchronously. I would like to allow them on multiple levels. Fetching multiple sources synchronously or asynchronously is already possible. I would like to have also then each tile separately in cache for re-use, but the system (`Tile`?) would have to know all the data it should render. 

So a possible scenario would be having (builtin) 
 - support for synchronized or separate multi-tile fetching (`getTileUrl` would return string for synchronous one request fetching - no changes, array to download mutliple tiles at once within multiple (or one?) image jobs)
 - support for multi-tile response
    - synchronized request - no changes, advanced data API has to handle it in a custom way, already possible
    - async request - either forced waiting for all data parts or replace missing one with black images and update when arrives, almost no change except for the missing data issue handling, since data can now be in general form
 - support for multi-tile caching - a Tile could have multiple caches so that it would be easy to drop or add data sources dynamically at runtime without the need to drop all the loaded data, as I do now when replacing the tiled image with `addTiledImage` with index specified
 - support for multi-tile rendering
   - the user would have to define how such data is rendered I guess (or maybe render as overlays if not specified) since they want to use it

I just need all the data at once when processing in WebGL. Having the option to say whether the data should be requested within one HTTP request or with multiple is just a sweet extension I would like to have. Processing the response of multiple tile data is the `TiledSource` responsibility (except for the possible forced synchronization or missing tile handling).

**Comment by iangilman** [comment]
This sounds like a lot of complexity that doesn't necessarily belong in the OSD core. I'd like to see how far you can go down the road of making it do what you need without any changes to OSD. If there's some element in OSD that needs an extra hook or something in order to enable some aspect, the hook could be added with the bulk of the implementation still happening outside of OSD. We can look at those things one at a time.

**Comment by Aiosa** [comment]
So I tried to design a solution, and what I came up with was: 

a) there should be flexibility in the data fetching, already implemented
b) there should be flexibility in the data caching, sort-of is, but still, it is not possible to control directly the cache creation
  - either create a new event fired when the cache is being created, or
  - add another method to `TileSource` that controls how the cache creation, instead of a hard-coded approach in the image job finisher after `tile-loaded` event
c) there should be flexibility in the data drawing
  - it is possible to use `tile-drawing` event to control what is drawn, but only with `context2D` property which disables cache
  - problematic are several other methods in the `Tile` class that rely on single data instance, and do not allow easy drawing control

Proposed solution that keeps the number of changes to the minium:

a) - none
b) delegate cache instance creation to the TileSource, I am not in favor of another event in this heavily used routine, and since TileSource already controls cache internals another method just first there perfectly
c) let TileSource control the Tile class it uses with a static property, i.e. OSD creates a new Tile instance as `new source.TileClass()` with `source.tileClass = $.Tile` as default.


So that the TileSource could now say which cache instances will be created, how they are created (already implemented), and which Tile class will be used within the system. Then, a plugin could simply implement new `MultiTile` class that extends `Tile` and assign it to a `TileSource`.

**Comment by Aiosa** [comment]
I was also thinking that, when already changing these, I could refactor the drawer as such, e.g. deprecate all drawing methods in `Tile`, move them to a Drawer instance (create a new `DrawerStrategy` class that implements the drawing, the `Drawer` would have attached a strategy class that handles the drawing, `useCanvas` would be deprecated and replaced with `drawingWith` string or class specifying which drawer to instantiate). But that's maybe too many changes at once. It would be doable if it involved only `Tile` and `Drawer` but a brief look told me there are probably other parts that account for HTML or Canvas drawing...

**Comment by iangilman** [comment]
Thank you for spelling it out like this! A few thoughts: 

* If we're going to upgrade the cache, we might as well make it as good as we can. I just want to make sure that it's backwards compatible and that the API continues to be easy for people who don't need to customize it.
* If I understand correctly, the key thing you want is a many-to-many relationship between tiles and the various bits of data they are composed of; is that right?  
* When it comes time to draw a frame, I think each tile should already be ready to draw. In the case of canvas, this means it should be something you can `drawImage` with, either an image or a canvas. It shouldn't be several images that get composited into each other at drawing time. In other words, no custom tile source code should be called during the actual drawing process; the tiles should already be ready. We need to keep this rule in place so the animation can be as smooth as possible. As soon as we allow custom code to run during frame drawing time, we are opening up to all sorts of performance degradation. 
* In light of that last point, do we need a cache system that can keep track of components and also the results of compositing components together? 
* If there are improvements we can make to the drawing system, for instance in preparation for WebGL, that sounds great, but it should be a separate patch from the cache improvements. We should think about which order makes sense. 

Those are my thoughts at the moment... Let's keep this conversation going!

**Comment by Aiosa** [comment]
I also don't like `context2D` property. It has been 'misused' heavily by various rendering-oriented plugins, but it removes the ability to cache, I would like to have a uniform data setter approach instead (e.g. attach three functions `getData[As...*]` that simply copy the cache API family approach) with a flag enabling or disabling cache use.

**Comment by iangilman** [comment]
Yes, being able to share the cache across multiple OSD instances would be great!

That's a good point about `tile-drawing`. I suppose to some extent it's unavoidable, but we should do everything we can to steer people toward processing ahead of time, including providing nice facilities to make that easy and documentation to warn against things that can reduce performance.

I agree that `context2D` isn't ideal… it kind of came in as a hack and has grown in importance since then. Coming up with a cleaner replacement would be wonderful. I do still think it needs to be canvas-ready pixels, so it can just be drawn immediately without any extra processing. There's no reason it shouldn't be cached, though.

Of course all of this (including the context2D replacement) should be implemented with backward compatibility in mind. Even if we deprecate some of the current systems, they should still function for the foreseeable future.

I suppose we have taken this discussion about as far as we can without getting into details… if you're up for it, please move forward with a patch and we can discuss further as the specifics come to light. As I mentioned before, if it's possible to do this in multiple chunks so it's not a single monster reworking, that would be lovely, but of course if it's not possible, we do what we've got to do :-)

**Comment by Aiosa** [comment]
Well, let's continue here: https://github.com/openseadragon/openseadragon/pull/2329

**Comment by iangilman** [comment]
Saying each Tile can only belong to a single TiledImage seems like a good rule. Of course, Tiles might share image data, but not share TiledImages. This is important, for instance, because it's the Tile that keeps track of its own positioning information.
